### Projet 1 
Groupe 3

Mathieu Muty - MUTM22090200

Gabrielle Walgraef - WALG13580100

Mamoudou Camara - CAMM12080206

In [1]:
# importation des libraries
import numpy as np
import pandas as pd

In [2]:
# chargement des fichiers csv
environnement = pd.read_csv('./data_projet/environnement.csv')
genre = pd.read_csv('./data_projet/genre.csv')
meteo = pd.read_csv('./data_projet/meteo.csv')
propulation = pd.read_csv('./data_projet/population.csv')
surface = pd.read_csv('./data_projet/surface.csv')
rapport = pd.read_excel('./data_projet/rapports-accident-2021.xlsx')

## Question 1
Les fichiers suivants : environnement, genre, meteo, population ainsi que surface ne contiennent pas de valeur
aberrante puisque ces fichiers nous donne les codes pour mieux comprendre le fichier excel 'rapport-accident-2021'

Nous nous concentrons donc sur le fichier 'rapports-accident-2021.xlsx' afin de déceler les valeurs extrêmes 
ou aberrantes

In [3]:
print(rapport.isna().sum())

print("type de donnée de la colonne CD_GENRE_ACCDN :", rapport['CD_GENRE_ACCDN'].dtype)
print("type de donnée de la colonne CD_GENRE_ACCDN :", rapport['MRC'].dtype)
print("nombre de valeur du dataframe : ", len(rapport))
# CD_COND_METEO meteo -> imputer nan by autre (code 99)
# CD_ENVRN_ACCDN environnement -> imputer nan by non précisé (code 0)
# CD_GENRE_ACCDN genre -> impossible de le savoir -> fillna('Unknown')(question 5 utilise cette column)
# CD_ETAT_SURFC surface -> imputer nan by autre (code 99)
# CD_ECLRM eclairage -> impossible de le savoir (non utilisé par les questions) - 
# VITESSE_AUTOR -> impossible de le savoir (non utilisé par les questions)

DT_ACCDN                          0
HR_ACCDN                          0
GENRE_CONDUCTEUR                  0
AGE_CONDUCTEUR                    0
gravite                           0
VALEURS_DOMMAGES                  0
NB_MORTS                          0
NB_BLESSES_GRAVES                 0
NB_BLESSES_LEGERS                 0
NB_VEH_IMPLIQUES_ACCDN            0
NB_DECES_PIETON                   0
NB_BLESSES_PIETON                 0
NB_VICTIMES_PIETON                0
NB_DECES_MOTO                     0
NB_BLESSES_MOTO                   0
NB_VICTIMES_MOTO                  0
NB_DECES_VELO                     0
NB_BLESSES_VELO                   0
NB_VICTIMES_VELO                  0
REG_ADM                          15
MRC                              15
VITESSE_AUTOR                  9310
CD_GENRE_ACCDN                  798
CD_ETAT_SURFC                  1163
CD_ECLRM                       1415
CD_ENVRN_ACCDN                 1239
CD_COND_METEO                  1352
nb_automobile_camion_leger  

## Démarche par rapport aux valeurs manquantes
Après identification des colonnes utilisées dans les prochaines questions et qui contiennent des valeurs manquantes,
Nous avons remarqué que ce cas de figure se retrouve dans la question 5 et la question 6

#### Cas de la question 5 :
Dans la question 5 nous utilisons la colonne **'CD_GENRE_ACCDN'** qui contient **798** valeurs manquantes.
Cette colonne nous indique un code **numérique** qui correspond au genre de l'accident.

Après étude de la question 5, qui incorpore des pourcentages a réaliser, nous avons décidé de supprimer ces valeurs manquantes pour le traitement de cette question même si ce nombre n'est pas significatif par rapport à la taille du dataFrame.


#### Cas de la question 6:
Dans la question 6 nous utilisons la colone **MRC** qui contient **15** valeurs manquantes. 
Après étude de la question 6, nous avons décidé de les supprimer afin de créer le dataframe **agg_tx_acc** sans erreur potentielle, de plus la suppression de ces valeurs n'a aucune incidence sur la question puisque nous devons sélectionner uniquement les 5 **MRC** avec le plus haut taux d'accident global.

## Question 2

In [17]:
#On met les dates de congés dans conges feries
conges_feries = [
    "2021-01-01",
    "2021-04-02",
    "2021-05-24",
    "2021-06-24",
    "2021-07-01",
    "2021-09-06",
    "2021-10-11",
    "2021-12-25"
]
#Conversion de la colonne de date au format datetime pour avoir AAAA-MM-JJ
rapport['DT_ACCDN'] = pd.to_datetime(rapport['DT_ACCDN'], format='%Y-%m-%d')

#Initialisons la variable EstCongéFérié comme si il correspond a aucune date feriée
rapport['EstCongéFérié'] = False

#On parcourt chaque ligne de la DataFrame pour voir une correspondance avec les jours feriés
#Puis dans le cas ou il y'a une correspondace, on met EstCongéFerié a True 
for index, row in rapport.iterrows():
    date_accident = row['DT_ACCDN'].strftime('%Y-%m-%d')
    if date_accident in conges_feries:
        rapport.at[index, 'EstCongéFérié'] = True
rapport[rapport['EstCongéFérié'] == True]

DT_ACCDN           HR_ACCDN GENRE_CONDUCTEUR  AGE_CONDUCTEUR  \
150    2021-05-24  12:00:00-12:59:00                H              31   
192    2021-06-24  13:00:00-13:59:00                H              56   
207    2021-07-01  05:00:00-05:59:00                F              19   
208    2021-07-01  10:00:00-10:59:00                F              29   
209    2021-07-01  12:00:00-12:59:00                F              68   
...           ...                ...              ...             ...   
100164 2021-07-01  16:00:00-16:59:00                F              63   
100269 2021-09-06  15:00:00-15:59:00                H              45   
100468 2021-12-25  02:00:00-02:59:00                H              24   
100469 2021-12-25  14:00:00-14:59:00                F              37   
100490 2021-12-25  15:00:00-15:59:00                H              19   

                                                  gravite  VALEURS_DOMMAGES  \
150                          Dommages matériels seulement      10068.018863   
192     Dommages matériels inférieurs au seuil de rapp...       2764.079837   
207                          Dommages matériels seulement       9844.460065   
208                                                 Léger       9213.429737   
209     Dommages matériels inférieurs au seuil de rapp...       2191.412262   
...                                                   ...               ...   
100164  Dommages matériels inférieurs au seuil de rapp...       3138.237014   
100269  Dommages matériels inférieurs au seuil de rapp...       3074.957276   
100468                       Dommages matériels seulement       9911.888705   
100469                       Dommages matériels seulement      10058.105165   
100490                       Dommages matériels seulement       9704.736827   

        NB_MORTS  NB_BLESSES_GRAVES  NB_BLESSES_LEGERS  \
150            0                  0                  0   
192            0                  0                  0   
207            0                  0                  0   
208            0                  0                  2   
209            0                  0                  0   
...          ...                ...                ...   
100164         0                  0                  0   
100269         0                  0                  0   
100468         0                  0                  0   
100469         0                  0                  0   
100490         0                  0                  0   

        NB_VEH_IMPLIQUES_ACCDN  ...  nb_cyclomoteur  nb_motocyclette  nb_taxi  \
150                          2  ...               0                0        0   
192                          2  ...               0                0        0   
207                          1  ...               0                0        0   
208                          3  ...               0                0        0   
209                          2  ...               0                0        0   
...                        ...  ...             ...              ...      ...   
100164                       1  ...               0                1        0   
100269                       2  ...               0                0        0   
100468                       1  ...               0                0        0   
100469                       1  ...               0                0        0   
100490                       1  ...               0                0        0   

        nb_urgence  nb_motoneige  nb_VHR  nb_autres_types  nb_veh_non_precise  \
150              0             0       0                0                   0   
192              0             0       0                0                   0   
207              0             0       0                0                   0   
208              0             0       0                0                   0   
209              0             0       0                0                   0   
...            ...         

## Question 3

In [19]:
rapport['EstWeekEnd'] = (rapport['DT_ACCDN'].dt.weekday >= 5)
rapport['EstWeekEnd']

0          True
1         False
2          True
3         False
4         False
          ...  
100514    False
100515    False
100516    False
100517     True
100518     True
Name: EstWeekEnd, Length: 100519, dtype: bool

## Question 4

Ici il n'y a pas de tri fait sur les valeurs manquantes car les analyses sont effectuées sur le total des accidents en fonction de leurs nombres de morts. Comme vérifié à la question 1, aucune valeurs ne sont manquantes dans les données utilisées.

In [71]:
"""
Fonction pour calculer du taux de mortalité en fonction du nombre d'accident et 
du total d'accident arrondit à la 2ème décimale
"""

def calcul_mortality_tx(nb_accdn, total_accdn):
    tx = (nb_accdn / total_accdn) * 100
    return round(tx, 2)

"""
mask afin de récupérer uniquement les accidents avec des morts issue d'un accident de voiture uniquement.
Autrement dit, ici, nous récupérons uniquement les accidents qui sont mortel avec des victimes qui ne sont
ni des piétons, ni des motards, ni des cyclistes.
"""
mask = (rapport['gravite'] == 'Mortel') & (rapport['NB_DECES_PIETON'] == 0) & (rapport['NB_DECES_MOTO'] == 0) & (rapport['NB_DECES_VELO'] == 0)

# Récupération du total d'accident par type d'utilisateur de la route impliqué
total_accdn = {
    "Véhicule" : len(rapport),
    "Piéton": rapport['NB_VICTIMES_PIETON'].sum(),
    "Moto": rapport['NB_VICTIMES_MOTO'].sum(),
    "Cycliste": rapport['NB_VICTIMES_VELO'].sum()
}

# création du premier dictionnaire afin de créer les catégories demandé et de leur assigner leurs valeurs
nb_accdn_mortel = {
    "Véhicule" : rapport[mask]['NB_MORTS'].sum(),
    "Piéton": rapport['NB_DECES_PIETON'].sum(),
    "Moto": rapport['NB_DECES_MOTO'].sum(),
    "Cycliste": rapport['NB_DECES_VELO'].sum()
}

# création du deuxième dictionnaire afin de caluler les taux de mortalités
tx_accdn_mortel = {
    "Véhicule": str(calcul_mortality_tx(nb_accdn_mortel['Véhicule'], total_accdn['Véhicule'])) + '%',
    "Piéton": str(calcul_mortality_tx(nb_accdn_mortel['Piéton'], total_accdn['Piéton'])) + '%',
    "Moto": str(calcul_mortality_tx(nb_accdn_mortel['Moto'], total_accdn['Moto'])) + '%',
    "Cycliste": str(calcul_mortality_tx(nb_accdn_mortel['Cycliste'], total_accdn['Cycliste'])) + '%'
}

tx_serie = pd.Series(tx_accdn_mortel)
tx_mortalité = pd.DataFrame({'Taux de mortalité': tx_serie})
tx_mortalité

Taux de mortalité
Véhicule             0.21%
Piéton               2.52%
Moto                 3.85%
Cycliste              1.1%

## Question 5

In [26]:
# a appliquer pour la question 5 : 
mask = rapport['CD_GENRE_ACCDN'].isna()
clean_report = rapport.drop(rapport[mask].index, axis=0)
print(clean_report.isna().sum())

DT_ACCDN                          0
HR_ACCDN                          0
GENRE_CONDUCTEUR                  0
AGE_CONDUCTEUR                    0
gravite                           0
VALEURS_DOMMAGES                  0
NB_MORTS                          0
NB_BLESSES_GRAVES                 0
NB_BLESSES_LEGERS                 0
NB_VEH_IMPLIQUES_ACCDN            0
NB_DECES_PIETON                   0
NB_BLESSES_PIETON                 0
NB_VICTIMES_PIETON                0
NB_DECES_MOTO                     0
NB_BLESSES_MOTO                   0
NB_VICTIMES_MOTO                  0
NB_DECES_VELO                     0
NB_BLESSES_VELO                   0
NB_VICTIMES_VELO                  0
REG_ADM                           7
MRC                               7
VITESSE_AUTOR                  9164
CD_GENRE_ACCDN                    0
CD_ETAT_SURFC                  1047
CD_ECLRM                       1303
CD_ENVRN_ACCDN                 1127
CD_COND_METEO                  1223
nb_automobile_camion_leger  

## Question 6
Dans le contexte de cette question, plusieurs valeurs étaient manquantes. Comme précisé dans la question 1, nous avons décidé de les supprimer du dataset

Calculez le taux d'accidents par nombre d'habitant dans chacune des MRC (Nombre
d'accident/Population). 

Produisez un nouvel objet de type DataFrame nommé agg_tx_acc qui présentera ce taux par gravité d'incident pour les 5 MRCs avec le plus haut taux d'accident global.

Le pourcentage indiqué doit être arrondi à deux (2) décimales et présenté avec le symbole "%".

Votre tableau doit contenir une colonne qui contient le total de toutes les lignes. Cette colonne doit se
nommer 'Total'.

In [35]:
# suppression des lignes où l'on retrouve des valeurs manquantes dans le cadre de la question 6
clean_mrc_mask = rapport['MRC'].isna()
clean_report = rapport.drop(rapport[clean_mrc_mask].index, axis=0)
print(clean_report.isna().sum())

DT_ACCDN                          0
HR_ACCDN                          0
GENRE_CONDUCTEUR                  0
AGE_CONDUCTEUR                    0
gravite                           0
VALEURS_DOMMAGES                  0
NB_MORTS                          0
NB_BLESSES_GRAVES                 0
NB_BLESSES_LEGERS                 0
NB_VEH_IMPLIQUES_ACCDN            0
NB_DECES_PIETON                   0
NB_BLESSES_PIETON                 0
NB_VICTIMES_PIETON                0
NB_DECES_MOTO                     0
NB_BLESSES_MOTO                   0
NB_VICTIMES_MOTO                  0
NB_DECES_VELO                     0
NB_BLESSES_VELO                   0
NB_VICTIMES_VELO                  0
REG_ADM                           0
MRC                               0
VITESSE_AUTOR                  9302
CD_GENRE_ACCDN                  790
CD_ETAT_SURFC                  1154
CD_ECLRM                       1406
CD_ENVRN_ACCDN                 1231
CD_COND_METEO                  1343
nb_automobile_camion_leger  

## Question 7

Créez un nouvel objet de type Series nommé temps qui présentera le maximum d'heures
écoulées entre chacun des accidents selon le mois de l'année

In [88]:
# Series pandas pour stocker chaque interval maximum par mois
temps = pd.Series({
    'january': 0,
    'february': 0,
    'march': 0,
    'april': 0,
    'may': 0,
    'june': 0,
    'july': 0,
    'august': 0,
    'september': 0,
    'october': 0,
    'november': 0,
    'december': 0
})

"""
Après quelques test sur le dataframe "rapport" en vue de traiter cette question, 
des valeurs 'Non précisé' ont été retrouvé dans la colonne 'HR_ACCDN'
Afin de ne pas générer de faux intervals, nous avons décider de supprimer ces valeurs.
"""
mask = rapport['HR_ACCDN'] == 'Non précisé'
clean_rapport = rapport.drop(rapport[mask].index, axis=0)


"""
ici, pour chaque mois, on récupère 
"""
for i in range(len(temps)):
    jun_sorted = clean_rapport[pd.DatetimeIndex(clean_rapport['DT_ACCDN']).month == (i + 1)].sort_values(by='DT_ACCDN')
    jun_sorted['StartedTime'] = jun_sorted['HR_ACCDN'].str.split('-').str[0]
    jun_sorted['date'] = jun_sorted['DT_ACCDN'].astype(str) + ' ' + jun_sorted['StartedTime']
    jun_sorted['date'] = pd.to_datetime(jun_sorted['date'], format='%Y-%m-%d %H:%M:%S')
    jun_sorted.sort_values('date', inplace=True)
    jun_sorted['time_difference'] = jun_sorted['date'].diff()
    max_interval_row = jun_sorted.loc[jun_sorted['time_difference'].idxmax()]
    temps[i] = max_interval_row['time_difference']

temps

january      0 days 05:00:00
february     0 days 04:00:00
march        0 days 04:00:00
april        0 days 04:00:00
may          0 days 06:00:00
june         0 days 02:00:00
july         0 days 02:00:00
august       0 days 02:00:00
september    0 days 02:00:00
october      0 days 02:00:00
november     0 days 02:00:00
december     0 days 03:00:00
dtype: object

## Question 8